{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "name": "T_Bio_Practice_3.ipynb",
      "provenance": [],
      "collapsed_sections": [],
      "authorship_tag": "ABX9TyNIETHiXLRwVgNFnoq4V/KL",
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/cappelchi/T-Bio/blob/master/T_Bio_Practice_3.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "PjQkF5wTXF54",
        "colab_type": "text"
      },
      "source": [
        "Hands-on Assignment: Transcriptomics 1\n",
        "YOUR ASSIGNMENT:\n",
        "\n",
        "**Run the RNA-seq pipeline and upload an excel file with the KRAS gene expression visualized as a bar plot.**"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "KjuU5iIoXb0f",
        "colab_type": "text"
      },
      "source": [
        "RNA-Seq\n",
        "Also known as whole transcriptome sequencing allows for the user to identify the RNA (mRNA generally) in a biological sample at any given time. This can be used to analyze the ever changing cellular transcriptome.\n",
        "\n",
        "Example: Jabbari, et al. used RNA-seq to investigate psoriasis and find new genes for functional analysis.  They compared their RNA-seq data to published array studies and found 1700 new candiadates. These were validated by qPCR, and comparison to functional databases for psoriasis supported their role in pathogenesis.\n",
        "\n",
        "The three different types of tumors are listed below.\n",
        "\n",
        "○ Estrogen Receptor positive (ER+),\n",
        "\n",
        "○ Human Epidermal Growth Factor positive (HER2+), and\n",
        "\n",
        "○ Triple Negative (TN), which is negative for expression of ER, HER2,\n",
        "and progesterone receptor.\n",
        "\n",
        "There are seven ER+ samples, thirteen TN samples. Each type of breast cancer\n",
        "has different characteristics, however, for this first practice exercise, we will\n",
        "not make any prior assumptions about the different types of breast cancer;\n",
        "we will look at all of the samples as a single group and see if a very basic\n",
        "RNA-seq analysis can show some differences between them. This is similar to\n",
        "an approach taken if you did not have any hypothesis as to whether the\n",
        "groups you are looking at are different. For example, this is the approach that\n",
        "would be taken by an oncologist with several unknown cancer biopsies who\n",
        "wants to see if these cancers were biologically different and therefore\n",
        "amenable to different treatments. (Data analysis steps like this are followed\n",
        "by biological analysis to confirm the differences between variation found in\n",
        "data and interpret them in biological terms.)\n",
        "\n",
        "<img src=\"https://raw.githubusercontent.com/cappelchi/T-Bio/master/practice3/RNA-Seq1.jpg\" width=\"600\" height=\"300\" />\n",
        "\n",
        "For this analysis, we will use an assembled genome as a scaffold (that is, a\n",
        "structure on which we hang sequences) on which to map the RNA reads. An\n",
        "assembled genome sequence is a chromosomal DNA sequence. It’s created\n",
        "by fragmenting the genome, sequencing the fragments, then putting the\n",
        "sequenced fragments together in the same order as in the real biological\n",
        "genome. The Human Genome Project is an example of an assembled\n",
        "genome. In this case, since this dataset is from a study of human breast\n",
        "cancer in mouse models, the assembled genome used as the reference\n",
        "genome is “human mouse,” a mixed assembled genome of Homo sapiens\n",
        "and Mus musculus."
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "YCeYCfTQdKnr",
        "colab_type": "text"
      },
      "source": [
        "## **Pipeline:**"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "2kxXAk32CYIj",
        "colab_type": "text"
      },
      "source": [
        "### Description:\n",
        "\n",
        "Step: no Pre-processing raw reads\n",
        "\n",
        "Step 1: Mapping on a reference genome\n",
        "\n",
        "Step 2: Calculating the abundance of reads aligned to each genomic element (i.e. exon, gene or isoform)\n",
        "\n",
        "Step 3: Biological Interpretation of gene expression profiles\n",
        "\n",
        "<img src=\"https://raw.githubusercontent.com/cappelchi/T-Bio/master/practice3/Pipeline1_tr1_p3.png\" width=\"600\" height=\"300\" />\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "5fAaghHgDNgw",
        "colab_type": "text"
      },
      "source": [
        "### Step 1: Mapping on a reference genome"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "F3_A9_IhaBPN",
        "colab_type": "text"
      },
      "source": [
        "#### Bowtie2\n",
        "\n",
        "Bowtie2  is a fast alignment algorithm that is based on the “seed” (or k-mer) approach. “Seed” substrings from the read and their reverse complements are extracted and aligned to the reference in an ungapped fashion. Then their positions on the reference are recorded and they are extended into full alignments using SIMD-accelerated dynamic programming.\n",
        "\n",
        "<img src=\"https://raw.githubusercontent.com/cappelchi/T-Bio/master/practice3/bowtie2-2.jpg\" width=\"600\" height=\"300\" />"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "NO5nnzl2SXvH",
        "colab_type": "text"
      },
      "source": [
        "### Step 2: Calculating the abundance of reads aligned to each genomic element (i.e. exon, gene or isoform)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "R7TwsGrlY7N9",
        "colab_type": "text"
      },
      "source": [
        "#### RSEM \n",
        "(RNA Seq using Expectation Maximization). RSEM is a software package for quantifying gene and isoform abundances from single-end or paired-end RNA-Seq data. RSEM outputs abundance estimates, 95% credibility intervals, and visualization files and can also simulate RNA-Seq data. In contrast to other existing tools, the software does not require a reference genome. Thus, in combination with a de novo transcriptome assembler, RSEM enables accurate transcript quantification for species without sequenced genomes. On simulated and real data sets, RSEM has superior or comparable performance to quantification methods that rely on a reference genome. Taking advantage of RSEM's ability to effectively use ambiguously-mapping reads, we show that accurate gene-level abundance estimates are best obtained with large numbers of short single-end reads. On the other hand, estimates of the relative frequencies of isoforms within single genes may be improved through the use of paired-end reads, depending on the number of possible splice forms for each gene.\n",
        "\n",
        "<img src=\"https://raw.githubusercontent.com/cappelchi/T-Bio/master/practice3/RSEM.png\" width=\"500\" height=\"270\">"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "E_2NjR24kXtb",
        "colab_type": "text"
      },
      "source": [
        "### Step 3: Biological Interpretation of gene expression profiles"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "ZZS7rl5no2qo",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "mypath = '/content/pipeline1p1ch3/'"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "1ucQk6rRTbMk",
        "colab_type": "code",
        "outputId": "1b51470c-1a16-4160-c23f-e111e9ec267d",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 426
        }
      },
      "source": [
        "!wget https://raw.githubusercontent.com/cappelchi/T-Bio/master/practice3/Practice_p1ch3_AllResultsRSEMRun.zip\n",
        "!unzip Practice_p1ch3_AllResultsRSEMRun.zip -d {mypath}"
      ],
      "execution_count": 2,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "--2020-03-06 08:12:55--  https://raw.githubusercontent.com/cappelchi/T-Bio/master/practice3/Practice_p1ch3_AllResultsRSEMRun.zip\n",
            "Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 151.101.0.133, 151.101.64.133, 151.101.128.133, ...\n",
            "Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|151.101.0.133|:443... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 3258926 (3.1M) [application/zip]\n",
            "Saving to: ‘Practice_p1ch3_AllResultsRSEMRun.zip’\n",
            "\n",
            "Practice_p1ch3_AllR 100%[===================>]   3.11M  13.6MB/s    in 0.2s    \n",
            "\n",
            "2020-03-06 08:13:01 (13.6 MB/s) - ‘Practice_p1ch3_AllResultsRSEMRun.zip’ saved [3258926/3258926]\n",
            "\n",
            "Archive:  Practice_p1ch3_AllResultsRSEMRun.zip\n",
            "  inflating: /content/pipeline1p1ch3/expression_genes_FPKM_not_filtered.txt  \n",
            "  inflating: /content/pipeline1p1ch3/expression_genes_FPKM.txt  \n",
            "  inflating: /content/pipeline1p1ch3/expression_genes_FPKM_with_annotation.txt  \n",
            "  inflating: /content/pipeline1p1ch3/expression_genes_not_filtered.txt  \n",
            "  inflating: /content/pipeline1p1ch3/expression_genes.txt  \n",
            "  inflating: /content/pipeline1p1ch3/expression_isoforms_FPKM_not_filtered.txt  \n",
            "  inflating: /content/pipeline1p1ch3/expression_isoforms_FPKM.txt  \n",
            "  inflating: /content/pipeline1p1ch3/expression_isoforms_FPKM_with_annotation.txt  \n",
            "  inflating: /content/pipeline1p1ch3/expression_isoforms_not_filtered.txt  \n",
            "  inflating: /content/pipeline1p1ch3/expression_isoforms.txt  \n",
            "  inflating: /content/pipeline1p1ch3/temp.txt  \n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "NCsOCiJbcFHn",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "#!pip install jupyter-datatables\n",
        "#!pip install pandas-summary"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "BNNtZeMqmwnV",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "import pandas as pd\n",
        "#from jupyter_datatables import init_datatables_mode\n",
        "#init_datatables_mode()  # initialize [DataTables]\n",
        "#from pandas_summary import DataFrameSummary\n",
        "from os import listdir\n",
        "from os.path import isfile, join"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "v0t5Z15poqn_",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "results = [f for f in listdir(mypath) if isfile(join(mypath, f))]"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "W1ZhXHxC1iQb",
        "colab_type": "code",
        "outputId": "9beba519-66e5-44f9-97e5-2e776d5bcd16",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 1000
        }
      },
      "source": [
        "for result in results:\n",
        "    if result[:11] == 'expression_':\n",
        "        df_name = result[11:-4]\n",
        "    else:\n",
        "        df_name = result[:-4]\n",
        "    comand = f'{df_name} = pd.read_csv(\"{mypath + result}\", sep = \"\\s+\")'\n",
        "    print (comand)\n",
        "    exec(comand)\n",
        "    print (df_name)\n",
        "    exec(f'print ({df_name}.describe())')"
      ],
      "execution_count": 13,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "genes_FPKM = pd.read_csv(\"/content/pipeline1p1ch3/expression_genes_FPKM.txt\", sep = \"\\s+\")\n",
            "genes_FPKM\n",
            "       ER-ERR1084763_PE  ER-ERR1084764_PE  ...  TN-ERR1084809_PE  TN-ERR1084810_PE\n",
            "count       3070.000000       3070.000000  ...       3070.000000       3070.000000\n",
            "mean         291.171003        295.306749  ...        365.988137        345.710003\n",
            "std         4768.355100       4653.392001  ...       5272.297673       5286.727094\n",
            "min            0.000000          0.000000  ...          0.000000          0.000000\n",
            "25%            0.000000          0.000000  ...          0.000000          0.000000\n",
            "50%            0.000000          0.000000  ...          0.000000          0.000000\n",
            "75%            0.000000          0.000000  ...          4.535000          3.232500\n",
            "max       183643.520000     183262.460000  ...     181583.440000     212177.950000\n",
            "\n",
            "[8 rows x 20 columns]\n",
            "temp = pd.read_csv(\"/content/pipeline1p1ch3/temp.txt\", sep = \"\\s+\")\n",
            "temp\n",
            "       ER-ERR1084763_PE  ER-ERR1084764_PE  ...  TN-ERR1084809_PE  TN-ERR1084810_PE\n",
            "count       4770.000000       4770.000000  ...       4770.000000       4770.000000\n",
            "mean         187.399197        190.061134  ...        235.551738        222.500069\n",
            "std         2814.827416       2864.443264  ...       3853.572237       3160.258741\n",
            "min            0.000000          0.000000  ...          0.000000          0.000000\n",
            "25%            0.000000          0.000000  ...          0.000000          0.000000\n",
            "50%            0.000000          0.000000  ...          0.000000          0.000000\n",
            "75%            0.000000          0.000000  ...          4.792500          3.760000\n",
            "max       101097.500000     109051.460000  ...     181449.890000     159585.810000\n",
            "\n",
            "[8 rows x 20 columns]\n",
            "genes = pd.read_csv(\"/content/pipeline1p1ch3/expression_genes.txt\", sep = \"\\s+\")\n",
            "genes\n",
            "       ER-ERR1084763_PE  ER-ERR1084764_PE  ...  TN-ERR1084809_PE  TN-ERR1084810_PE\n",
            "count       3327.000000       3327.000000  ...       3327.000000       3327.000000\n",
            "mean          40.496441         42.055077  ...         74.794857         88.770520\n",
            "std          774.413367        782.818784  ...       1022.189693       1092.705131\n",
            "min            0.000000          0.000000  ...          0.000000          0.000000\n",
            "25%            0.000000          0.000000  ...          0.000000          0.000000\n",
            "50%            0.000000          0.000000  ...          0.000000          0.000000\n",
            "75%            0.000000          0.000000  ...          1.060000          1.000000\n",
            "max        38645.550000      38398.030000  ...      42695.520000      40359.670000\n",
            "\n",
            "[8 rows x 20 columns]\n",
            "genes_FPKM_not_filtered = pd.read_csv(\"/content/pipeline1p1ch3/expression_genes_FPKM_not_filtered.txt\", sep = \"\\s+\")\n",
            "genes_FPKM_not_filtered\n",
            "       ER-ERR1084763_PE  ER-ERR1084764_PE  ...  TN-ERR1084809_PE  TN-ERR1084810_PE\n",
            "count      96697.000000      96697.000000  ...      96697.000000      96697.000000\n",
            "mean           9.244373          9.375649  ...         11.619909         10.976343\n",
            "std          851.031422        830.633029  ...        941.467299        943.796979\n",
            "min            0.000000          0.000000  ...          0.000000          0.000000\n",
            "25%            0.000000          0.000000  ...          0.000000          0.000000\n",
            "50%            0.000000          0.000000  ...          0.000000          0.000000\n",
            "75%            0.000000          0.000000  ...          0.000000          0.000000\n",
            "max       183643.520000     183262.460000  ...     181583.440000     212177.950000\n",
            "\n",
            "[8 rows x 20 columns]\n",
            "isoforms_FPKM_with_annotation = pd.read_csv(\"/content/pipeline1p1ch3/expression_isoforms_FPKM_with_annotation.txt\", sep = \"\\s+\")\n",
            "isoforms_FPKM_with_annotation\n",
            "       ER-ERR1084763_PE  ER-ERR1084764_PE  ...  TN-ERR1084809_PE  TN-ERR1084810_PE\n",
            "count       4770.000000       4770.000000  ...       4770.000000       4770.000000\n",
            "mean         187.399197        190.061134  ...        235.551738        222.500069\n",
            "std         2814.827416       2864.443264  ...       3853.572237       3160.258741\n",
            "min            0.000000          0.000000  ...          0.000000          0.000000\n",
            "25%            0.000000          0.000000  ...          0.000000          0.000000\n",
            "50%            0.000000          0.000000  ...          0.000000          0.000000\n",
            "75%            0.000000          0.000000  ...          4.792500          3.760000\n",
            "max       101097.500000     109051.460000  ...     181449.890000     159585.810000\n",
            "\n",
            "[8 rows x 20 columns]\n",
            "isoforms_not_filtered = pd.read_csv(\"/content/pipeline1p1ch3/expression_isoforms_not_filtered.txt\", sep = \"\\s+\")\n",
            "isoforms_not_filtered\n",
            "       ER-ERR1084763_PE  ER-ERR1084764_PE  ...  TN-ERR1084809_PE  TN-ERR1084810_PE\n",
            "count     290899.000000     290899.000000  ...     290899.000000     290899.000000\n",
            "mean           0.463157          0.480985  ...          0.855451          1.015291\n",
            "std           61.071798         64.539016  ...         95.389742         88.164138\n",
            "min            0.000000          0.000000  ...          0.000000          0.000000\n",
            "25%            0.000000          0.000000  ...          0.000000          0.000000\n",
            "50%            0.000000          0.000000  ...          0.000000          0.000000\n",
            "75%            0.000000          0.000000  ...          0.000000          0.000000\n",
            "max        23195.240000      26322.130000  ...      34669.680000      24297.610000\n",
            "\n",
            "[8 rows x 20 columns]\n",
            "isoforms = pd.read_csv(\"/content/pipeline1p1ch3/expression_isoforms.txt\", sep = \"\\s+\")\n",
            "isoforms\n",
            "       ER-ERR1084763_PE  ER-ERR1084764_PE  ...  TN-ERR1084809_PE  TN-ERR1084810_PE\n",
            "count       5005.000000       5005.000000  ...       5005.000000       5005.000000\n",
            "mean          26.919223         27.955093  ...         49.717371         59.007810\n",
            "std          464.876791        491.296944  ...        725.627342        669.657311\n",
            "min            0.000000          0.000000  ...          0.000000          0.000000\n",
            "25%            0.000000          0.000000  ...          0.000000          0.000000\n",
            "50%            0.000000          0.000000  ...          0.000000          0.000000\n",
            "75%            0.000000          0.000000  ...          1.160000          1.010000\n",
            "max        23195.240000      26322.130000  ...      34669.680000      24297.610000\n",
            "\n",
            "[8 rows x 20 columns]\n",
            "isoforms_FPKM = pd.read_csv(\"/content/pipeline1p1ch3/expression_isoforms_FPKM.txt\", sep = \"\\s+\")\n",
            "isoforms_FPKM\n",
            "       ER-ERR1084763_PE  ER-ERR1084764_PE  ...  TN-ERR1084809_PE  TN-ERR1084810_PE\n",
            "count       4770.000000       4770.000000  ...       4770.000000       4770.000000\n",
            "mean         187.399197        190.061134  ...        235.551738        222.500069\n",
            "std         2814.827416       2864.443264  ...       3853.572237       3160.258741\n",
            "min            0.000000          0.000000  ...          0.000000          0.000000\n",
            "25%            0.000000          0.000000  ...          0.000000          0.000000\n",
            "50%            0.000000          0.000000  ...          0.000000          0.000000\n",
            "75%            0.000000          0.000000  ...          4.792500          3.760000\n",
            "max       101097.500000     109051.460000  ...     181449.890000     159585.810000\n",
            "\n",
            "[8 rows x 20 columns]\n",
            "isoforms_FPKM_not_filtered = pd.read_csv(\"/content/pipeline1p1ch3/expression_isoforms_FPKM_not_filtered.txt\", sep = \"\\s+\")\n",
            "isoforms_FPKM_not_filtered\n",
            "       ER-ERR1084763_PE  ER-ERR1084764_PE  ...  TN-ERR1084809_PE  TN-ERR1084810_PE\n",
            "count     290899.000000     290899.000000  ...     290899.000000     290899.000000\n",
            "mean           3.072898          3.116536  ...          3.862545          3.648619\n",
            "std          361.193471        367.554728  ...        494.314707        405.622764\n",
            "min            0.000000          0.000000  ...          0.000000          0.000000\n",
            "25%            0.000000          0.000000  ...          0.000000          0.000000\n",
            "50%            0.000000          0.000000  ...          0.000000          0.000000\n",
            "75%            0.000000          0.000000  ...          0.000000          0.000000\n",
            "max       101097.500000     109051.460000  ...     181449.890000     159585.810000\n",
            "\n",
            "[8 rows x 20 columns]\n",
            "genes_not_filtered = pd.read_csv(\"/content/pipeline1p1ch3/expression_genes_not_filtered.txt\", sep = \"\\s+\")\n",
            "genes_not_filtered\n",
            "       ER-ERR1084763_PE  ER-ERR1084764_PE  ...  TN-ERR1084809_PE  TN-ERR1084810_PE\n",
            "count      96697.000000      96697.000000  ...      96697.000000      96697.000000\n",
            "mean           1.393342          1.446973  ...          2.573502          3.054355\n",
            "std          143.814441        145.386002  ...        190.067743        203.301083\n",
            "min            0.000000          0.000000  ...          0.000000          0.000000\n",
            "25%            0.000000          0.000000  ...          0.000000          0.000000\n",
            "50%            0.000000          0.000000  ...          0.000000          0.000000\n",
            "75%            0.000000          0.000000  ...          0.000000          0.000000\n",
            "max        38645.550000      38398.030000  ...      42695.520000      40359.670000\n",
            "\n",
            "[8 rows x 20 columns]\n",
            "genes_FPKM_with_annotation = pd.read_csv(\"/content/pipeline1p1ch3/expression_genes_FPKM_with_annotation.txt\", sep = \"\\s+\")\n",
            "genes_FPKM_with_annotation\n",
            "       ER-ERR1084763_PE  ER-ERR1084764_PE  ...  TN-ERR1084809_PE  TN-ERR1084810_PE\n",
            "count       3070.000000       3070.000000  ...       3070.000000       3070.000000\n",
            "mean         291.171003        295.306749  ...        365.988137        345.710003\n",
            "std         4768.355100       4653.392001  ...       5272.297673       5286.727094\n",
            "min            0.000000          0.000000  ...          0.000000          0.000000\n",
            "25%            0.000000          0.000000  ...          0.000000          0.000000\n",
            "50%            0.000000          0.000000  ...          0.000000          0.000000\n",
            "75%            0.000000          0.000000  ...          4.535000          3.232500\n",
            "max       183643.520000     183262.460000  ...     181583.440000     212177.950000\n",
            "\n",
            "[8 rows x 20 columns]\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "4n5-DWc2-0u3",
        "colab_type": "text"
      },
      "source": [
        "The output files labeled with ‘FPKM’ contain expression measurements in\n",
        "FPKM, whereas the others simply contain the count of reads per gene (or\n",
        "isoform). The files labeled with ‘_not_filtred.txt’ contain genes with zeroexpression for all samples whereas those genes are filtered out in the other files.\n",
        "Recall from section 2.1 that Bowtie can output tables of the expression levels of all genes (including those with expression levels of zero), or a table of only those genes with expression above zero. For most applications, we will only be interested in those genes with expression above zero. Further, it is computationally much more expensive, when dealing with large datasets, to open a table including zero-expression-level genes. Therefore, we will practice using the table of only those genes with expression; in other words, a table in which genes with expression=0 have been removed.\n",
        "Open the files in Excel (Open the gene expression file to look at genes,and the isoform expression file to look at isoforms). Click on the links to the gene and isoform expression tables to open these.\n",
        "Row 1 contains the names of the samples. Column A contains the names of\n",
        "each gene in the GTF file for which there was some expression above zero in any sample. Remember, the values represent the level of expression for the gene (named in Column A), not values for individual reads. (We have now moved from the level of reads to the level of genes.)\n",
        "Searching in the interactive table below, a subset of 20 genes, for the Ensembl name ENSG00000133703 will locate KRAS, a known proto-oncogene. You will further explore the biological significance of your\n",
        "results in the following courses.\n",
        "The isoform expression table looks just like the gene expression table, except that there are more rows, since some genes have more than one isoform. For the rest of this practice example, we will use the gene expression table."
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "4Rz0zqBKAsY0",
        "colab_type": "text"
      },
      "source": [
        "Выходные файлы, помеченные «FPKM», содержат измерения экспрессии в FPKM, в то время как другие просто содержат количество чтений на ген (или изоформу). Файлы, помеченные «_not_filtred.txt», содержат гены с нулевой экспрессией для всех образцов, тогда как эти гены отфильтрованы в других файлах. Напомним из раздела 2.1, что Bowtie может выводить таблицы уровней экспрессии всех генов (в том числе с уровнями экспрессии ноль) или таблицу только тех генов с экспрессией выше нуля. Для большинства приложений нас будут интересовать только те гены с выражением выше нуля. Кроме того, в вычислительном отношении гораздо дороже при работе с большими наборами данных открывать таблицу, включающую гены с нулевым уровнем экспрессии. Поэтому мы будем практиковать использование таблицы только тех генов с выражением; другими словами, таблица, в которой были удалены гены с выражением = 0. Откройте файлы в Excel (откройте файл генного выражения, чтобы посмотреть на гены, и файл выражения изоформы, чтобы посмотреть на изоформы). Нажмите на ссылки на таблицы экспрессии генов и изоформ, чтобы открыть их. Строка 1 содержит названия образцов. Столбец A содержит названия каждого гена в файле GTF, для которого в любом образце было некоторое выражение выше нуля. Помните, что значения представляют уровень экспрессии для гена (назван в столбце A), а не значения для отдельных операций чтения. (Теперь мы перешли от уровня чтения к уровню генов.) Поиск в интерактивной таблице ниже подмножества из 20 генов для имени Ensembl ENSG00000133703 позволит найти KRAS, известный протоонкоген. Далее вы изучите биологическую значимость ваших результатов на следующих курсах. Таблица экспрессии изоформ выглядит так же, как таблица экспрессии генов, за исключением того, что имеется больше строк, поскольку некоторые гены имеют более одной изоформы. В оставшейся части этого практического примера мы будем использовать таблицу экспрессии генов."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "kS49wzuA-z9R",
        "colab_type": "code",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 457
        },
        "outputId": "dff074b8-6455-4260-e90c-a7a938d47bfc"
      },
      "source": [
        "genes_FPKM"
      ],
      "execution_count": 14,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>id</th>\n",
              "      <th>ER-ERR1084763_PE</th>\n",
              "      <th>ER-ERR1084764_PE</th>\n",
              "      <th>ER-ERR1084765_PE</th>\n",
              "      <th>ER-ERR1084775_PE</th>\n",
              "      <th>ER-ERR1084805_PE</th>\n",
              "      <th>ER-ERR1084806_PE</th>\n",
              "      <th>ER-ERR1084811_PE</th>\n",
              "      <th>TN-ERR1084766_PE</th>\n",
              "      <th>TN-ERR1084768_PE</th>\n",
              "      <th>TN-ERR1084798_PE</th>\n",
              "      <th>TN-ERR1084799_PE</th>\n",
              "      <th>TN-ERR1084800_PE</th>\n",
              "      <th>TN-ERR1084801_PE</th>\n",
              "      <th>TN-ERR1084802_PE</th>\n",
              "      <th>TN-ERR1084803_PE</th>\n",
              "      <th>TN-ERR1084804_PE</th>\n",
              "      <th>TN-ERR1084807_PE</th>\n",
              "      <th>TN-ERR1084808_PE</th>\n",
              "      <th>TN-ERR1084809_PE</th>\n",
              "      <th>TN-ERR1084810_PE</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>ENSG00000001630</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>12.95</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>10.27</td>\n",
              "      <td>10.28</td>\n",
              "      <td>0.00</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>ENSG00000001631</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>3.49</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>ENSG00000002834</td>\n",
              "      <td>13707.91</td>\n",
              "      <td>15284.44</td>\n",
              "      <td>15155.14</td>\n",
              "      <td>8173.33</td>\n",
              "      <td>7621.02</td>\n",
              "      <td>21836.52</td>\n",
              "      <td>20287.39</td>\n",
              "      <td>4810.31</td>\n",
              "      <td>794.38</td>\n",
              "      <td>11611.88</td>\n",
              "      <td>8155.97</td>\n",
              "      <td>5737.05</td>\n",
              "      <td>5914.49</td>\n",
              "      <td>7506.39</td>\n",
              "      <td>11705.55</td>\n",
              "      <td>3982.07</td>\n",
              "      <td>5158.16</td>\n",
              "      <td>5831.16</td>\n",
              "      <td>2946.15</td>\n",
              "      <td>9858.35</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>ENSG00000003402</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.54</td>\n",
              "      <td>0.41</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>2.53</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.44</td>\n",
              "      <td>1.23</td>\n",
              "      <td>0.26</td>\n",
              "      <td>1.83</td>\n",
              "      <td>0.70</td>\n",
              "      <td>3.40</td>\n",
              "      <td>0.00</td>\n",
              "      <td>1.04</td>\n",
              "      <td>0.92</td>\n",
              "      <td>2.09</td>\n",
              "      <td>2.38</td>\n",
              "      <td>4.39</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>ENSG00000003436</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.39</td>\n",
              "      <td>0.53</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>1.61</td>\n",
              "      <td>0.50</td>\n",
              "      <td>0.00</td>\n",
              "      <td>1.03</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.51</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>1.08</td>\n",
              "      <td>0.00</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3065</th>\n",
              "      <td>ENSMUSG00000097705</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>46.67</td>\n",
              "      <td>83.16</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>2.76</td>\n",
              "      <td>38.81</td>\n",
              "      <td>33.89</td>\n",
              "      <td>8.31</td>\n",
              "      <td>73.94</td>\n",
              "      <td>0.00</td>\n",
              "      <td>46.09</td>\n",
              "      <td>47.86</td>\n",
              "      <td>20.06</td>\n",
              "      <td>319.77</td>\n",
              "      <td>0.00</td>\n",
              "      <td>50.12</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3066</th>\n",
              "      <td>ENSMUSG00000097779</td>\n",
              "      <td>5.87</td>\n",
              "      <td>5.55</td>\n",
              "      <td>18.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>3.08</td>\n",
              "      <td>2.87</td>\n",
              "      <td>0.00</td>\n",
              "      <td>73.65</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>11.70</td>\n",
              "      <td>0.00</td>\n",
              "      <td>22.51</td>\n",
              "      <td>5.41</td>\n",
              "      <td>2.98</td>\n",
              "      <td>14.38</td>\n",
              "      <td>3.46</td>\n",
              "      <td>443.57</td>\n",
              "      <td>40.66</td>\n",
              "      <td>5.18</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3067</th>\n",
              "      <td>ENSMUSG00000097873</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>3.43</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>1.43</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>2.63</td>\n",
              "      <td>0.00</td>\n",
              "      <td>1.05</td>\n",
              "      <td>1.60</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.86</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3068</th>\n",
              "      <td>ENSMUSG00000098650</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>49.60</td>\n",
              "      <td>782.79</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>62.99</td>\n",
              "      <td>23.33</td>\n",
              "      <td>124.08</td>\n",
              "      <td>23.47</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>12.99</td>\n",
              "      <td>241.46</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3069</th>\n",
              "      <td>ENSMUSG00000099115</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>5.59</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>3070 rows × 21 columns</p>\n",
              "</div>"
            ],
            "text/plain": [
              "                      id  ER-ERR1084763_PE  ...  TN-ERR1084809_PE  TN-ERR1084810_PE\n",
              "0        ENSG00000001630              0.00  ...             10.28              0.00\n",
              "1        ENSG00000001631              0.00  ...              0.00              0.00\n",
              "2        ENSG00000002834          13707.91  ...           2946.15           9858.35\n",
              "3        ENSG00000003402              0.00  ...              2.38              4.39\n",
              "4        ENSG00000003436              0.00  ...              1.08              0.00\n",
              "...                  ...               ...  ...               ...               ...\n",
              "3065  ENSMUSG00000097705              0.00  ...              0.00             50.12\n",
              "3066  ENSMUSG00000097779              5.87  ...             40.66              5.18\n",
              "3067  ENSMUSG00000097873              0.00  ...              0.00              0.86\n",
              "3068  ENSMUSG00000098650              0.00  ...             12.99            241.46\n",
              "3069  ENSMUSG00000099115              0.00  ...              0.00              0.00\n",
              "\n",
              "[3070 rows x 21 columns]"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 14
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "Zx3JnN2rBmFH",
        "colab_type": "code",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 457
        },
        "outputId": "cee720f6-d5a4-4bc4-f1a1-a51baa98c431"
      },
      "source": [
        "genes_FPKM_with_annotation"
      ],
      "execution_count": 15,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>id</th>\n",
              "      <th>ER-ERR1084763_PE</th>\n",
              "      <th>ER-ERR1084764_PE</th>\n",
              "      <th>ER-ERR1084765_PE</th>\n",
              "      <th>ER-ERR1084775_PE</th>\n",
              "      <th>ER-ERR1084805_PE</th>\n",
              "      <th>ER-ERR1084806_PE</th>\n",
              "      <th>ER-ERR1084811_PE</th>\n",
              "      <th>TN-ERR1084766_PE</th>\n",
              "      <th>TN-ERR1084768_PE</th>\n",
              "      <th>TN-ERR1084798_PE</th>\n",
              "      <th>TN-ERR1084799_PE</th>\n",
              "      <th>TN-ERR1084800_PE</th>\n",
              "      <th>TN-ERR1084801_PE</th>\n",
              "      <th>TN-ERR1084802_PE</th>\n",
              "      <th>TN-ERR1084803_PE</th>\n",
              "      <th>TN-ERR1084804_PE</th>\n",
              "      <th>TN-ERR1084807_PE</th>\n",
              "      <th>TN-ERR1084808_PE</th>\n",
              "      <th>TN-ERR1084809_PE</th>\n",
              "      <th>TN-ERR1084810_PE</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>ENSG00000001630_ensembl_havana_protein_coding</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>12.95</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>10.27</td>\n",
              "      <td>10.28</td>\n",
              "      <td>0.00</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>ENSG00000001631_KRIT1_ensembl_havana</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>3.49</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>ENSG00000002834_ensembl_havana_protein_coding</td>\n",
              "      <td>13707.91</td>\n",
              "      <td>15284.44</td>\n",
              "      <td>15155.14</td>\n",
              "      <td>8173.33</td>\n",
              "      <td>7621.02</td>\n",
              "      <td>21836.52</td>\n",
              "      <td>20287.39</td>\n",
              "      <td>4810.31</td>\n",
              "      <td>794.38</td>\n",
              "      <td>11611.88</td>\n",
              "      <td>8155.97</td>\n",
              "      <td>5737.05</td>\n",
              "      <td>5914.49</td>\n",
              "      <td>7506.39</td>\n",
              "      <td>11705.55</td>\n",
              "      <td>3982.07</td>\n",
              "      <td>5158.16</td>\n",
              "      <td>5831.16</td>\n",
              "      <td>2946.15</td>\n",
              "      <td>9858.35</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>ENSG00000003402_CFLAR_ensembl_havana</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.54</td>\n",
              "      <td>0.41</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>2.53</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.44</td>\n",
              "      <td>1.23</td>\n",
              "      <td>0.26</td>\n",
              "      <td>1.83</td>\n",
              "      <td>0.70</td>\n",
              "      <td>3.40</td>\n",
              "      <td>0.00</td>\n",
              "      <td>1.04</td>\n",
              "      <td>0.92</td>\n",
              "      <td>2.09</td>\n",
              "      <td>2.38</td>\n",
              "      <td>4.39</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>ENSG00000003436_ensembl_havana_protein_coding</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.39</td>\n",
              "      <td>0.53</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>1.61</td>\n",
              "      <td>0.50</td>\n",
              "      <td>0.00</td>\n",
              "      <td>1.03</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.51</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>1.08</td>\n",
              "      <td>0.00</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3065</th>\n",
              "      <td>ENSMUSG00000097705_Gm26740_havana</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>46.67</td>\n",
              "      <td>83.16</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>2.76</td>\n",
              "      <td>38.81</td>\n",
              "      <td>33.89</td>\n",
              "      <td>8.31</td>\n",
              "      <td>73.94</td>\n",
              "      <td>0.00</td>\n",
              "      <td>46.09</td>\n",
              "      <td>47.86</td>\n",
              "      <td>20.06</td>\n",
              "      <td>319.77</td>\n",
              "      <td>0.00</td>\n",
              "      <td>50.12</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3066</th>\n",
              "      <td>ENSMUSG00000097779_4833407H14Rik_ensembl</td>\n",
              "      <td>5.87</td>\n",
              "      <td>5.55</td>\n",
              "      <td>18.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>3.08</td>\n",
              "      <td>2.87</td>\n",
              "      <td>0.00</td>\n",
              "      <td>73.65</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>11.70</td>\n",
              "      <td>0.00</td>\n",
              "      <td>22.51</td>\n",
              "      <td>5.41</td>\n",
              "      <td>2.98</td>\n",
              "      <td>14.38</td>\n",
              "      <td>3.46</td>\n",
              "      <td>443.57</td>\n",
              "      <td>40.66</td>\n",
              "      <td>5.18</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3067</th>\n",
              "      <td>ENSMUSG00000097873_Gm26867_ensembl</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>3.43</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>1.43</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>2.63</td>\n",
              "      <td>0.00</td>\n",
              "      <td>1.05</td>\n",
              "      <td>1.60</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.86</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3068</th>\n",
              "      <td>ENSMUSG00000098650_havana_protein_coding</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>49.60</td>\n",
              "      <td>782.79</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>62.99</td>\n",
              "      <td>23.33</td>\n",
              "      <td>124.08</td>\n",
              "      <td>23.47</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>12.99</td>\n",
              "      <td>241.46</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3069</th>\n",
              "      <td>ENSMUSG00000099115_Gm17190_havana</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>5.59</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>3070 rows × 21 columns</p>\n",
              "</div>"
            ],
            "text/plain": [
              "                                                 id  ...  TN-ERR1084810_PE\n",
              "0     ENSG00000001630_ensembl_havana_protein_coding  ...              0.00\n",
              "1              ENSG00000001631_KRIT1_ensembl_havana  ...              0.00\n",
              "2     ENSG00000002834_ensembl_havana_protein_coding  ...           9858.35\n",
              "3              ENSG00000003402_CFLAR_ensembl_havana  ...              4.39\n",
              "4     ENSG00000003436_ensembl_havana_protein_coding  ...              0.00\n",
              "...                                             ...  ...               ...\n",
              "3065              ENSMUSG00000097705_Gm26740_havana  ...             50.12\n",
              "3066       ENSMUSG00000097779_4833407H14Rik_ensembl  ...              5.18\n",
              "3067             ENSMUSG00000097873_Gm26867_ensembl  ...              0.86\n",
              "3068       ENSMUSG00000098650_havana_protein_coding  ...            241.46\n",
              "3069              ENSMUSG00000099115_Gm17190_havana  ...              0.00\n",
              "\n",
              "[3070 rows x 21 columns]"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 15
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "c8ieIMEsBy20",
        "colab_type": "code",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 457
        },
        "outputId": "b4cadc98-8541-4424-817e-a9934a113300"
      },
      "source": [
        "isoforms_FPKM"
      ],
      "execution_count": 16,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>id</th>\n",
              "      <th>ER-ERR1084763_PE</th>\n",
              "      <th>ER-ERR1084764_PE</th>\n",
              "      <th>ER-ERR1084765_PE</th>\n",
              "      <th>ER-ERR1084775_PE</th>\n",
              "      <th>ER-ERR1084805_PE</th>\n",
              "      <th>ER-ERR1084806_PE</th>\n",
              "      <th>ER-ERR1084811_PE</th>\n",
              "      <th>TN-ERR1084766_PE</th>\n",
              "      <th>TN-ERR1084768_PE</th>\n",
              "      <th>TN-ERR1084798_PE</th>\n",
              "      <th>TN-ERR1084799_PE</th>\n",
              "      <th>TN-ERR1084800_PE</th>\n",
              "      <th>TN-ERR1084801_PE</th>\n",
              "      <th>TN-ERR1084802_PE</th>\n",
              "      <th>TN-ERR1084803_PE</th>\n",
              "      <th>TN-ERR1084804_PE</th>\n",
              "      <th>TN-ERR1084807_PE</th>\n",
              "      <th>TN-ERR1084808_PE</th>\n",
              "      <th>TN-ERR1084809_PE</th>\n",
              "      <th>TN-ERR1084810_PE</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>ENSMUST00000000412</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>1.44</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>ENSMUST00000000704</td>\n",
              "      <td>0.00</td>\n",
              "      <td>5.34</td>\n",
              "      <td>0.00</td>\n",
              "      <td>8.88</td>\n",
              "      <td>0.00</td>\n",
              "      <td>63.53</td>\n",
              "      <td>11.11</td>\n",
              "      <td>8.17</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>8.43</td>\n",
              "      <td>11.03</td>\n",
              "      <td>0.00</td>\n",
              "      <td>15.50</td>\n",
              "      <td>2.87</td>\n",
              "      <td>8.30</td>\n",
              "      <td>23.30</td>\n",
              "      <td>113.32</td>\n",
              "      <td>3.01</td>\n",
              "      <td>4.99</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>ENSMUST00000001675</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>6.78</td>\n",
              "      <td>2.58</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>12.84</td>\n",
              "      <td>4.90</td>\n",
              "      <td>0.00</td>\n",
              "      <td>2.23</td>\n",
              "      <td>1.66</td>\n",
              "      <td>1.61</td>\n",
              "      <td>0.83</td>\n",
              "      <td>3.07</td>\n",
              "      <td>0.00</td>\n",
              "      <td>1.63</td>\n",
              "      <td>0.00</td>\n",
              "      <td>6.30</td>\n",
              "      <td>0.00</td>\n",
              "      <td>2.96</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>ENSMUST00000001706</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>3.38</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>ENSMUST00000001809</td>\n",
              "      <td>32.65</td>\n",
              "      <td>59.09</td>\n",
              "      <td>91.26</td>\n",
              "      <td>37.08</td>\n",
              "      <td>34.55</td>\n",
              "      <td>10.76</td>\n",
              "      <td>63.32</td>\n",
              "      <td>54.12</td>\n",
              "      <td>15.13</td>\n",
              "      <td>30.46</td>\n",
              "      <td>40.34</td>\n",
              "      <td>32.67</td>\n",
              "      <td>23.78</td>\n",
              "      <td>114.72</td>\n",
              "      <td>18.52</td>\n",
              "      <td>43.75</td>\n",
              "      <td>22.17</td>\n",
              "      <td>83.85</td>\n",
              "      <td>36.09</td>\n",
              "      <td>45.72</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4765</th>\n",
              "      <td>ENST00000610020</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.51</td>\n",
              "      <td>0.00</td>\n",
              "      <td>1.57</td>\n",
              "      <td>0.24</td>\n",
              "      <td>3.52</td>\n",
              "      <td>1.37</td>\n",
              "      <td>1.22</td>\n",
              "      <td>2.73</td>\n",
              "      <td>3.14</td>\n",
              "      <td>3.48</td>\n",
              "      <td>97.96</td>\n",
              "      <td>2.48</td>\n",
              "      <td>6.61</td>\n",
              "      <td>1.29</td>\n",
              "      <td>0.78</td>\n",
              "      <td>3.42</td>\n",
              "      <td>0.72</td>\n",
              "      <td>6.66</td>\n",
              "      <td>1.69</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4766</th>\n",
              "      <td>ENST00000610026</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>1.19</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4767</th>\n",
              "      <td>ENST00000610067</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>10.47</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4768</th>\n",
              "      <td>ENST00000610091</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>2.99</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>1.18</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.64</td>\n",
              "      <td>0.51</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4769</th>\n",
              "      <td>ENST00000610279</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>2.93</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>3.76</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>4770 rows × 21 columns</p>\n",
              "</div>"
            ],
            "text/plain": [
              "                      id  ER-ERR1084763_PE  ...  TN-ERR1084809_PE  TN-ERR1084810_PE\n",
              "0     ENSMUST00000000412              0.00  ...              0.00              0.00\n",
              "1     ENSMUST00000000704              0.00  ...              3.01              4.99\n",
              "2     ENSMUST00000001675              0.00  ...              0.00              2.96\n",
              "3     ENSMUST00000001706              0.00  ...              0.00              0.00\n",
              "4     ENSMUST00000001809             32.65  ...             36.09             45.72\n",
              "...                  ...               ...  ...               ...               ...\n",
              "4765     ENST00000610020              0.00  ...              6.66              1.69\n",
              "4766     ENST00000610026              0.00  ...              0.00              0.00\n",
              "4767     ENST00000610067              0.00  ...              0.00              0.00\n",
              "4768     ENST00000610091              0.00  ...              0.64              0.51\n",
              "4769     ENST00000610279              0.00  ...              0.00              0.00\n",
              "\n",
              "[4770 rows x 21 columns]"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 16
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "PiNv4d1yB80g",
        "colab_type": "code",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 457
        },
        "outputId": "10e38624-d111-4204-841d-2418d83ea583"
      },
      "source": [
        "isoforms_FPKM_with_annotation"
      ],
      "execution_count": 18,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>id</th>\n",
              "      <th>ER-ERR1084763_PE</th>\n",
              "      <th>ER-ERR1084764_PE</th>\n",
              "      <th>ER-ERR1084765_PE</th>\n",
              "      <th>ER-ERR1084775_PE</th>\n",
              "      <th>ER-ERR1084805_PE</th>\n",
              "      <th>ER-ERR1084806_PE</th>\n",
              "      <th>ER-ERR1084811_PE</th>\n",
              "      <th>TN-ERR1084766_PE</th>\n",
              "      <th>TN-ERR1084768_PE</th>\n",
              "      <th>TN-ERR1084798_PE</th>\n",
              "      <th>TN-ERR1084799_PE</th>\n",
              "      <th>TN-ERR1084800_PE</th>\n",
              "      <th>TN-ERR1084801_PE</th>\n",
              "      <th>TN-ERR1084802_PE</th>\n",
              "      <th>TN-ERR1084803_PE</th>\n",
              "      <th>TN-ERR1084804_PE</th>\n",
              "      <th>TN-ERR1084807_PE</th>\n",
              "      <th>TN-ERR1084808_PE</th>\n",
              "      <th>TN-ERR1084809_PE</th>\n",
              "      <th>TN-ERR1084810_PE</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>ENSMUST00000000412_ensembl_havana_protein_coding</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>1.44</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>ENSMUST00000000704_ensembl_havana_protein_coding</td>\n",
              "      <td>0.00</td>\n",
              "      <td>5.34</td>\n",
              "      <td>0.00</td>\n",
              "      <td>8.88</td>\n",
              "      <td>0.00</td>\n",
              "      <td>63.53</td>\n",
              "      <td>11.11</td>\n",
              "      <td>8.17</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>8.43</td>\n",
              "      <td>11.03</td>\n",
              "      <td>0.00</td>\n",
              "      <td>15.50</td>\n",
              "      <td>2.87</td>\n",
              "      <td>8.30</td>\n",
              "      <td>23.30</td>\n",
              "      <td>113.32</td>\n",
              "      <td>3.01</td>\n",
              "      <td>4.99</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>ENSMUST00000001675_ensembl_havana_protein_coding</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>6.78</td>\n",
              "      <td>2.58</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>12.84</td>\n",
              "      <td>4.90</td>\n",
              "      <td>0.00</td>\n",
              "      <td>2.23</td>\n",
              "      <td>1.66</td>\n",
              "      <td>1.61</td>\n",
              "      <td>0.83</td>\n",
              "      <td>3.07</td>\n",
              "      <td>0.00</td>\n",
              "      <td>1.63</td>\n",
              "      <td>0.00</td>\n",
              "      <td>6.30</td>\n",
              "      <td>0.00</td>\n",
              "      <td>2.96</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>ENSMUST00000001706_ensembl_havana_protein_coding</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>3.38</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>ENSMUST00000001809_ensembl_havana_protein_coding</td>\n",
              "      <td>32.65</td>\n",
              "      <td>59.09</td>\n",
              "      <td>91.26</td>\n",
              "      <td>37.08</td>\n",
              "      <td>34.55</td>\n",
              "      <td>10.76</td>\n",
              "      <td>63.32</td>\n",
              "      <td>54.12</td>\n",
              "      <td>15.13</td>\n",
              "      <td>30.46</td>\n",
              "      <td>40.34</td>\n",
              "      <td>32.67</td>\n",
              "      <td>23.78</td>\n",
              "      <td>114.72</td>\n",
              "      <td>18.52</td>\n",
              "      <td>43.75</td>\n",
              "      <td>22.17</td>\n",
              "      <td>83.85</td>\n",
              "      <td>36.09</td>\n",
              "      <td>45.72</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4765</th>\n",
              "      <td>ENST00000610020_ensembl_havana_protein_coding</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.51</td>\n",
              "      <td>0.00</td>\n",
              "      <td>1.57</td>\n",
              "      <td>0.24</td>\n",
              "      <td>3.52</td>\n",
              "      <td>1.37</td>\n",
              "      <td>1.22</td>\n",
              "      <td>2.73</td>\n",
              "      <td>3.14</td>\n",
              "      <td>3.48</td>\n",
              "      <td>97.96</td>\n",
              "      <td>2.48</td>\n",
              "      <td>6.61</td>\n",
              "      <td>1.29</td>\n",
              "      <td>0.78</td>\n",
              "      <td>3.42</td>\n",
              "      <td>0.72</td>\n",
              "      <td>6.66</td>\n",
              "      <td>1.69</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4766</th>\n",
              "      <td>ENST00000610026_PPA1_ensembl_havana</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>1.19</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4767</th>\n",
              "      <td>ENST00000610067_LINC01128_havana</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>10.47</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4768</th>\n",
              "      <td>ENST00000610091_RP11-417J8.6_ensembl_havana</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>2.99</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>1.18</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.64</td>\n",
              "      <td>0.51</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4769</th>\n",
              "      <td>ENST00000610279_RP11-57G10.8_havana</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>2.93</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>3.76</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>4770 rows × 21 columns</p>\n",
              "</div>"
            ],
            "text/plain": [
              "                                                    id  ...  TN-ERR1084810_PE\n",
              "0     ENSMUST00000000412_ensembl_havana_protein_coding  ...              0.00\n",
              "1     ENSMUST00000000704_ensembl_havana_protein_coding  ...              4.99\n",
              "2     ENSMUST00000001675_ensembl_havana_protein_coding  ...              2.96\n",
              "3     ENSMUST00000001706_ensembl_havana_protein_coding  ...              0.00\n",
              "4     ENSMUST00000001809_ensembl_havana_protein_coding  ...             45.72\n",
              "...                                                ...  ...               ...\n",
              "4765     ENST00000610020_ensembl_havana_protein_coding  ...              1.69\n",
              "4766               ENST00000610026_PPA1_ensembl_havana  ...              0.00\n",
              "4767                  ENST00000610067_LINC01128_havana  ...              0.00\n",
              "4768       ENST00000610091_RP11-417J8.6_ensembl_havana  ...              0.51\n",
              "4769               ENST00000610279_RP11-57G10.8_havana  ...              0.00\n",
              "\n",
              "[4770 rows x 21 columns]"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 18
        }
      ]
    }
  ]
}